### ISSUE TITLE
[GTA VC] [GTA III] Ultimate ASI Loader doesn't work on wine

### ISSUE THREAD

**Issue by thepoke32** [ThirteenAG/Ultimate-ASI-Loader#12]
Title says all.
I tried renaming dinput8 into xlive, ddraw, d3d9 etc.

**Comment by ThirteenAG** [comment]
https://github.com/ThirteenAG/WidescreenFixesPack/issues/207

**Comment by dariusworks** [comment]
should be mentioned like on site, so no need for this in future :pager:

**Comment by ThirteenAG** [comment]
I'm not sure how to phrase it since I never used that myself, so feel free to pull request updated readme if you want.

**Comment by keybreak** [comment]
@ThirteenAG 
Problem is - it still doesn't work in Wine, just tested with Win 5.7-staging under Manjaro Linux...

- Game itself works
- Default .asi loader works (judging that it loads from main game directory)
- CLEO works
- Anything from /scripts or d3d8to9 doesn't work at all (which is real pain, since it's only way to use reshade with ssrtgi...)

For me overrides of d3d9 / d3d8 / dinput8 haven't help like with case for Gun in this issue above.

Is there anything that can be done?

**Comment by ThirteenAG** [comment]
I have no idea, I never used wine.

**Comment by keybreak** [comment]
Well, i know...

Although, regardless of your own preferences for OS - i think you should give it a go, at least for fun if anything, since people are already playing stuff like GTA V / Far Cry 5 / Hitman 2 on either very close to Win or even better performance on Linux in some cases it's only matter of time before more and more people would want to replay classics with d3d8to9 or reshade or asiloader...

So they'll come here anyway, just like myself :smile: 

Thx for all your hard work, i've been a huge fan for years and years!

_________

If anyone else have some hacky / override solution specifically for GTA III / VC - please do post it! :+1:

**Comment by elishacloud** [comment]
@keybreak, I have not used Wine much either.  However, I have seen that API hooking does not work the same way on Linux as it does on Windows.  You may want to make sure that both `ForceEntryPointHook` and `DontLoadFromDLLMain` are disabled in the ini file.

Also, you could try using my [dxwrapper](https://github.com/elishacloud/dxwrapper) project.  It also has the ability to load ASI scripts and should work with Wine on Linux.

**Comment by keybreak** [comment]
@elishacloud 
Thx!

Well, at least in theory ideally it should work same as Windows (i'll try more different games though, something like Max Payne or Mafia)...

```
DontLoadFromDllMain=0
ForceEntryPointHook=0
```

Those two options sadly doesn't change anything...

Haven't heard of **dxwrapper** project before, looks a little bit more complex on configuration side, i've tried to copy those files "as is":
```
dinput8.dll
dxwrapper.dll
dxwrapper.ini
Stub.ini
```

Which doesn't seem to make difference either, it it correctly done, or should i try some settings?

**Comment by elishacloud** [comment]
Did you see a dxwrapper log file get generated?  Can you attach it?

Did you configure the `dxwrapper.ini` file?  These settings may need to be set if you want to load the ASI plug-ins:

```ini
[Plugins]
LoadPlugins                = 0
LoadFromScriptsOnly        = 0
```

Finally, try it without any ASI plug-ins first.  It could be the plug-in that has an issue, not the dxwrapper ASI loader.

**Comment by keybreak** [comment]
> Did you see a dxwrapper log file get generated?

Nope :frowning_face: 

> Did you configure the dxwrapper.ini file?

Not until now, i have applied settings you've just mentioned and also this one:
```
[Compatibility]
D3d8to9                    = 1
```
Since i need at least **D3d8to9** for ReShade to work, but obviously .asi as well, since without windescreenfix, silent's patch, SkyGFX - GTA III / VC is very sad story!

> Finally, try it without any ASI plug-ins first. It could be the plug-in that has an issue, not the dxwrapper ASI loader.

Makes sense, but no, just tested apparently it's not plugins...

**Comment by elishacloud** [comment]
If you did not see a log file then it means that the game is not loading the `dinput8.dll` file.  Try renaming it to some other supported dll.  Apparently the game is either loading that dll from the Windows folder or does not use that dll.

You could try these ones:
* dsound.dll
* d3d8.dll

**Comment by keybreak** [comment]
You got me thinking, since i know for fact that GTA III use this file, that must be some Windows folder "Wine" priority, so i did override of ***dinput8 (native, builtin)** in Wine...

Well - certainly d3d9 wrapper at least somehow worked, but game started at default 640x480 resolution, so when i've decided to change it to my native prior to further test - it came up as some jagged messed constant crashes + crash of my compositor, so i had to go to Linux terminal for killing that nasty process.

But at least i have log, maybe it will tell you something useful
[dxwrapper-gta3.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4589577/dxwrapper-gta3.log)

**Comment by elishacloud** [comment]
> "certainly d3d9 wrapper at least somehow worked"

Can you tell me what the `d3d8.dll` and `d3d9.dll` files are?  Is `d3d8.dll` from dxwrapper and `d3d9.dll` from ReShade?

My recommendation is first to disable everything in dxwrapper.ini and then see if that works.  Then enable the settings one-by-one to see which ones have issues.

BTW: the log looks ok other than it does not show the process ending, since it was terminated prematurely.

**Comment by elishacloud** [comment]
> "but unusable due to any mouse move cause freeze of whole game for something like 3-5 seconds, not using mouse works fine though...weird"

Mouse issues could have to do with `dinput8.dll`.  Most likely this is used for the mouse controls.

**Comment by keybreak** [comment]
**d3d8.dll** - leftover from widescreenfix, sorry forgot to remove it
**d3d9.dll** - yes, Reshade

So to be on safe side i've just clean installed with **dxwrapper**, first i've tried with only this options change from default config, just to see if asi loading from **scripts** folder works:
```
[Plugins]
LoadPlugins                = 1
LoadFromScriptsOnly        = 0
```
Tested with **SilentPatchIII.asi** which is always easy to see by headlights - nope, it doesn't work at all (no log, no effect etc).

If i use ***dinput8 (native, builtin)** override in Wine, there is a log which seems to indicate that everything is ok and .asi plugin is loaded:
[dxwrapper-gta3.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591388/dxwrapper-gta3.log)

But in fact it is not active (headlights obviously doesn't work). Mouse works fine though!
Well...at least this test tells us that with ***dinput8 (native, builtin)** Wine override it's actual **dinput8** from game folder is in use, otherwise **dxwrapper-gta3.log** won't be there, right?
________________

Then i've removed **dxwrapper-gta3.log** and added (also removed Wine's ***dinput8 (native, builtin)** override):
```
[Compatibility]
D3d8to9                    = 1
```

No log, hence no effect.

Then again, added ***dinput8 (native, builtin)** override in Wine:
[dxwrapper-gta3.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591462/dxwrapper-gta3.log)

It works and produces log, but again headlights doesn't work and it started with resolution reset to 640x480 - never a good sign. Mouse still works though!
Note that there are still d3d8 mentioned in log, perhaps one from Windows folder?

Also, here's log created by Wine (dxvk), maybe it will tell you something (specifically warnings)
[gta3_d3d9.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591480/gta3_d3d9.log)

If i add Reshade:
```
reshade-shaders/
d3d9.dll
d3dcompiler_47.dll
ReShade.ini
```

It certainly works and loads, but mouse is not usable. Logs looks kinda similar:
[dxwrapper-gta3.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591554/dxwrapper-gta3.log)
[gta3_d3d9.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591555/gta3_d3d9.log)
[d3d9.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4591556/d3d9.log)

And...btw headlights from silent are still not here, so nothing works fully :smiley: 

_________

However, when i add widescreen fix with it's **d3d8.dll** - headlights from silent's patch are working as well as widescreenfix!
Also it's certainly silent's headlights, since i've used:

GTA3.WidescreenFix.ini
```
[MAIN]
FixVehicleLights = 0
```

Mouse obviously still doesn't.

___________

So sine it's **d3d8.dll** which makes a big change, i figure i'd try again:
```
[Compatibility]
D3d8to9                    = 0
```

Yes, it works. And mouse is fine (since ReShade haven't been loaded due to dx8 in use)


_________


It seems that mouse bug is caused by combination of using **dinput8** dll override in Wine and using ReShade though. But it's more likely a problem of **dxwrapper** or widescreenfix **d3d8.dll**, than ReShade. since i've used this version of ReShade sucessfully with mutiple dx9, dx10 and dx11 games under Wine

**Comment by keybreak** [comment]
Yeah it should be wrapper or **d3d8.dll**, since even with Wine's ***dinput8 (native, builtin)** overrided @ThirteenAG Ultimate-ASI-Loader doesn't start game at all...
It says:
```
wine: Unhandled exception 0x40000015 in thread 9 at address 100435F2 (thread 0009), starting
```

**Comment by elishacloud** [comment]
From the log I would guess that the issue has to do with ReShade.  It also looks like ReShade is trying to hook OpenGL, which I don't think it should be doing for this game.

```text
13:07:22:504 [00009] | INFO  | Initializing crosire's ReShade version '4.6.1.826' (32-bit) built on '2020-04-14 19:08:11' loaded from "C:\Games\GTA III\d3d9.dll" into "C:\Games\GTA III\gta3.exe" ...
13:07:22:513 [00009] | INFO  | Registering hooks for "user32.dll" ...
13:07:22:513 [00009] | INFO  | > Libraries loaded.
13:07:22:513 [00009] | INFO  | > Found 13 match(es). Installing ...
13:07:22:516 [00009] | INFO  | Registering hooks for "ws2_32.dll" ...
13:07:22:516 [00009] | INFO  | > Libraries loaded.
13:07:22:516 [00009] | INFO  | > Found 8 match(es). Installing ...
13:07:22:519 [00009] | INFO  | Registering hooks for "C:\windows\system32\d2d1.dll" ...
13:07:22:520 [00009] | INFO  | > Delayed.
13:07:22:520 [00009] | INFO  | Registering hooks for "C:\windows\system32\d3d9.dll" ...
13:07:22:520 [00009] | INFO  | > Delayed until first call to an exported function.
13:07:22:520 [00009] | INFO  | Registering hooks for "C:\windows\system32\d3d10.dll" ...
13:07:22:521 [00009] | INFO  | > Delayed.
13:07:22:521 [00009] | INFO  | Registering hooks for "C:\windows\system32\d3d10_1.dll" ...
13:07:22:521 [00009] | INFO  | > Delayed.
13:07:22:521 [00009] | INFO  | Registering hooks for "C:\windows\system32\d3d11.dll" ...
13:07:22:522 [00009] | INFO  | > Delayed.
13:07:22:522 [00009] | INFO  | Registering hooks for "C:\windows\system32\d3d12.dll" ...
13:07:22:523 [00009] | INFO  | > Delayed.
13:07:22:523 [00009] | INFO  | Registering hooks for "C:\windows\system32\dxgi.dll" ...
13:07:22:523 [00009] | INFO  | > Delayed.
13:07:22:523 [00009] | INFO  | Registering hooks for "C:\windows\system32\opengl32.dll" ...
13:07:22:523 [00009] | INFO  | > Libraries loaded.
13:07:22:524 [00009] | INFO  | > Found 359 match(es). Installing ...
13:07:22:528 [00009] | INFO  | Initialized.
```

**Comment by keybreak** [comment]
>  It also looks like ReShade is trying to hook OpenGL, which I don't think it should be doing for this game.

Nah, it's some process ReShade does in log for any game regardless of actual api actually used, i don't know why but that's part of initialization process...

Here's for example how d3d9.log looks for Mirror's edge (which is dx9 and plays with ReShade successfully)
[d3d9.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4595171/d3d9.log)

But anyway, i don't think it's ReShade at all, since **Ultimate-ASI-Loader** or **dxwrapper** doesn't work without it's **d3d9.dll** at all, for example if you try to test something if i remove ReShade files and try last working **dxwrapper** config with something that require working D3d8to9 like SkyGFX for example - you got this error if you try to start new game or load save:
![image](https://user-images.githubusercontent.com/47599989/81335160-d5458100-90af-11ea-9968-425be0fc9103.png)

**Comment by elishacloud** [comment]
> since Ultimate-ASI-Loader or dxwrapper doesn't work without it's d3d9.dll at all

This means you are using the wrong dll for dxwrapper and Ultimate-ASI-Loader.  What is happening is ReShade is loading dxwrapper/Ultimate-ASI-Loader rather than the game loading it.  This can lead to all kinds of problems because it can mean that dxwrapper and Ultimate-ASI-Loader are loading too late in the game startup.

Here is what you should do:
1. Remove everything from the game and have nothing but the game only.
2. Then drop in dxwrapper and it's stub only, making sure that dxwrapper has all settings disabled.
3. Enable dxwrapper's settings one-by-one until they are all enabled and the game still works correctly.
4. Once dxwrapper is working then start adding other items (ReShade, ASI plug-ins, etc.) one-by-one, making sure not to replace or delete any of dxwrapper's files.

Note: I strongly recommend using d3d8.dll for the dxwrapper's stub.  Using dinput8.dll or some other dll may cause dxwrapper to load too late in the game loading process, which will cause all kinds of issues.

**Comment by keybreak** [comment]
Well i did that already and posted results it seems...

> Note: I strongly recommend using d3d8.dll for the dxwrapper's stub. Using dinput8.dll or some other dll may cause dxwrapper to load too late in the game loading process, which will cause all kinds of issues.

That's not an option, if you do that with override ***d3d8 (native, builtin)** it will throw exactly same unhandled exception on loading, if you don't do override obviously **dxwrapper** don't work at all.

**Comment by elishacloud** [comment]
> That's not an option, if you do that with override *d3d8 (native, builtin) it will throw exactly same unhandled exception on loading, if you don't do override obviously dxwrapper don't work at all.

If you enable `d3d8to9` it will also override the native, builtin d3d8 anyways.  The code hooks d3d8 and redirects it to d3d9 so the native d3d8 is bypassed.  Maybe just backup the native d3d8 and replace it with dxwrapper's stub and then enable `d3d8to9`.  The native d3d8 may not be needed if you are using `d3d8to9`.

**Comment by keybreak** [comment]
Ok, makes sense...But it doesn't work as well.

______

You know, i've just tested again to exclude all asi loaders hackery at all:
1. Clean Wine prefix 5.7-staging
2. Clean GTA III game (retail v1.0)
3. Latest https://github.com/crosire/d3d8to9 (d3d8.dll)
4. Latest ReShade

ReShade loads and technically works, but mouse movement freezes whole game for 3-5 seconds, so it is unusable.

I would assume it is problem of **d3d8to9** / Wine, @crosire should i create issue for d3d8to9?
Or maybe you can give some advice, if we missed something?

**Comment by elishacloud** [comment]
Can you zip up and attach the native d3d8.dll file the game uses?

**Comment by keybreak** [comment]
Sure, i suppose you mean Wine's ones since there are no native d3d8.dll in vanilla game?
[windows.zip](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4596218/windows.zip)

**Comment by elishacloud** [comment]
> there are no native d3d8.dll in vanilla game

I am a little bit confused.  Earlier you said `if you do that with override *d3d8 (native, builtin).`  Why are you worried about overriding the native, builtin d3d8 if there is no native d3d8 file in the game folder to override?

Does the game run natively without any updates or patches?  If so, you should be able to drop the d3d8.dll stub file only into the game's folder and it should continue running as is, albeit you will get a pop-up that says it cannot load dxwrapper.

**Comment by keybreak** [comment]
Well, turns out we both don't understand what exactly wine's overrides mean...
It's some voodoo magic for most cases it just wouldn't work without override.

- As i've said game itself doesn't come with d3d8.dll / dinput8.dll

- Natively it runs fine without any updates and patches

- If you add just d3d8.dll to game folder (assuming you mean dxwrapper's one) - it starts game with resolution reset to 640x480, which usually indicates some problems, but seems to run fine, i've done it couple of times. I'd assume it runs in dx9, since at least this log from DXVK appears:
[gta3_d3d9.log](https://github.com/ThirteenAG/Ultimate-ASI-Loader/files/4598095/gta3_d3d9.log)
But usually if you really run dx9 on Wine with dxvk there also should be **gta-vc.dxvk-cache**
So it most likely still in dx8 mode.

- If inside Wine you add an override ***d3d8 (native, builtin)** in that case, it will actually mean at least for case with GTA that it really do an attempt to run d3d8 from game directory, since on launch you will get this:
![image](https://user-images.githubusercontent.com/47599989/81391975-a9191700-9126-11ea-9119-c04d10005daa.png)

- If you add those to the mix:
```
dxwrapper.dll
dxwrapper.ini
Stub.ini
```
with just those parameters:
```
[Plugins]
LoadPlugins                = 1
LoadFromScriptsOnly        = 0
```
It will run fine (in dx8 mode i would assume), .asi plugins would also work and **dxwrapper-gta3.log** will be created.
If you try to remove Wine's ***d3d8 (native, builtin)** override - it won't work with .asi and won't create **dxwrapper-gta3.log.**

- If you add:
```
[Compatibility]
D3d8to9                    = 1
```
**gta3.dxvk-cache** finally gets created - which strongly indicates that you're running dx9, or at least trying. Game seems to work, .asi loading from scripts folder as well, but how to actually test in game?

Well, as far as i know - you can try few options one by one:
- SkyGfx
![image](https://user-images.githubusercontent.com/47599989/81393193-94d61980-9128-11ea-8b64-fe9a36f724e0.png) 

- Sharptrails
![image](https://user-images.githubusercontent.com/47599989/81393170-8ab41b00-9128-11ea-85a5-cfa81444cfe9.png)
![image](https://user-images.githubusercontent.com/47599989/81393263-bd5e1380-9128-11ea-8862-75051f1f13b4.png)

- ReShade
Well, you know the drill already - it certainly works since you wouldn't even see ReShade's overlay without dx9 - but problem with mouse persists.

**Comment by elishacloud** [comment]
> **gta3.dxvk-cache** finally gets created - which strongly indicates that you're running dx9, or at least trying. Game seems to work, .asi loading from scripts folder as well, but how to actually test in game?

If I understood your detailed response, It looks like the game works until you add ReShade.  There may be an issue with ReShade.  I have never seen this type of an issue before.  But if it works fine without ReShade then you could try using some other d3d9 shader (such as [injectSMAA](http://mrhaandi.blogspot.com/p/injectsmaa.html)) and see if that works.

**Comment by keybreak** [comment]
Well, as expected it won't even start game with both Ultimate ASI Loader and dxwrapper, but that's ok coz ReShade is only known injector which fully works on Wine / DXVK

**Comment by infoman** [comment]
Apparently it works when you also have **real** `dinput8` placed in `system32`/`syswow64` directory

**Comment by keybreak** [comment]
@infoman 
Hey!
Can you please be more specific?

I've tested on:
Manjaro KDE stable
Wine 5.7-staging
dxvk 1.7L
overrides: *d3d8 (native, builtin)

Ultimate ASI loader, still see same Unhandled exception with Windows 7 SP1 original dinput8 files thrown to system32 / system64, while trying SKYGFX for example.

**Comment by infoman** [comment]
I tried running a GTA III bundle built by another person, with what seems to be ASI Loader pretending to be dinput8.dll. First, when trying to just set dinput8 to native, I got unhandled exception errors and that's all. Then, I installed dinput8 from PlayOnLinux settings, and got some of the bundled mods working, but car models were broken. After also setting d3d8 to native, everything worked even without having real d3d8 installed.

Maybe if you are trying to replace d3d8 then you must have real d3d8 installed.
